cdap-ui/app/cdap/components/PluginJSONCreator/Create/Content/Filters/FilterCollection/FilterShowlistInput/index.tsx

```diff
@@ -72,78 +74,65 @@ const FilterShowlistInputView: React.FC<IFilterShowlistInputProps> = ({
+        .valueSeq()
+        .map((info) => fromJS(info).get('name'))
+        .filter((widgetName) => !isNil(widgetName) && !isNil(widgetName))
+    : List([]);
 
   function setShowProperty(showID: string, property: string) {
     return (val) => {
-      setShowToInfo((prevObjs) => ({
-        ...prevObjs,
-        [showID]: { ...prevObjs[showID], [property]: val },
-      }));
+      setShowToInfo(fromJS(showToInfo).setIn([showID, property], val));
     };
   }
 
-  function addShowToFilter(filterObjID: string, index?: number) {
+  function addShowToFilter(filterObjID: string, index: number) {
     const newShowID = 'Show_' + uuidV4();
 
-    setShowToInfo({
-      ...showToInfo,
-      [newShowID]: {
-        name: '',
-      } as IPropertyShowConfig,
-    });
-
-    if (index === undefined) {
-      setFilterToShowList({
-        ...filterToShowList,
-        [filterObjID]: [...filterToShowList[filterObjID], newShowID],
-      });
+    setShowToInfo(
+      showToInfo.set(
+        newShowID,
+        fromJS({
+          name: '',
+          type: '',
+        })
+      )
+    );
+
+    const showlist = filterToShowList.get(filterObjID);
+
+    let newShowlist;
+    if (showlist.isEmpty()) {
+      newShowlist = showlist.insert(0, newShowID);
     } else {
-      const showList = filterToShowList[filterObjID];
-
-      if (showList.length === 0) {
-        showList.splice(0, 0, newShowID);
-      } else {
-        showList.splice(index + 1, 0, newShowID);
-      }
-
-      setFilterToShowList({
-        ...filterToShowList,
-        [filterObjID]: showList,
-      });
+      newShowlist = showlist.insert(index + 1, newShowID);
     }
+
+    setFilterToShowList(filterToShowList.set(filterObjID, newShowlist));
   }
 
   function deleteShowFromFilter(filterObjID: string, index: number) {
-    const showList = filterToShowList[filterObjID];
-
-    const showToDelete = showList[index];
-
-    showList.splice(index, 1);
+    const showlist = filterToShowList.get(filterObjID);
+    const showToDelete = showlist.get(index);
 
-    setFilterToShowList({
-      ...filterToShowList,
-      [filterObjID]: showList,
-    });
+    const newShowlist = showlist.remove(index);
+    setFilterToShowList(filterToShowList.set(filterObjID, newShowlist));
 
-    const { [showToDelete]: tmp, ...restShowToInfo } = showToInfo;
-    setShowToInfo(restShowToInfo);
+    const newShowToInfo = fromJS(showToInfo.delete(showToDelete));
+    setShowToInfo(newShowToInfo);
   }
 
   return (
-    <If condition={!isNil(filterToShowList[filterID])}>
+    <If condition={filterToShowList.has(filterID)}>
       <Heading type={HeadingTypes.h6} label="Add widgets to configure" />
-      {filterToShowList[filterID].map((showID: string, showIndex: number) => {
+      {filterToShowList.get(filterID).map((showID: string, showIndex: number) => {
         return (
-          <If key={showID} condition={showToInfo[showID]}>
+          <If key={showID} condition={showToInfo.has(showID)}>
```

Comment:
missed this before, but if the top level of the row is an `<If>`, you can conditionally render the row by checking `if (!showToInfo.has(showID)) { return null; }` instead

Reply:
Resolved